src/mount.rs

```diff
@@ -0,0 +1,655 @@
+        };
+
+        if hunk_index >= self.hunks.len() {
+            None
+        } else {
+            Some(hunk_index)
+        }
+    }
+
+    pub fn find_hunk_for_file(&self, path: &Apath) -> Option<u32> {
+        self.find_hunk_index_for_file(path)
+            .map(|index| self.hunks[index].index)
+    }
+
+    pub fn find_hunks_for_subdir(&self, path: &Apath, recursive: bool) -> Vec<u32> {
+        /*
+         * Appending an empty string to the path allows us to search for the first file
+         * in the target directory. This is needed as a file and a directory with the same name are not
+         * stored in succession.
+         *
+         * Example (from the apath test):
+         *  - /b/a
+         *  - /b/b
+         *  - /b/c
+         *  - /b/a/c
+         *  - /b/b/c
+         */
+        let search_path = path.append("");
+        let directory_start_hunk = match self.find_hunk_index_for_file(&search_path) {
+            Some(index) => index,
+            None => return vec![],
+        };
+
+        let mut result = Vec::new();
+        result.push(self.hunks[directory_start_hunk].index);
+        for hunk in &self.hunks[directory_start_hunk + 1..] {
+            if !path.is_prefix_of(&hunk.start_path) {
+                break;
+            }
+
+            if !recursive && hunk.start_path[path.len() + 1..].contains('/') {
+                /* hunk does already contain directory content */
+                break;
+            }
+
+            /* hunk still contains subtree elements of that path */
+            result.push(hunk.index);
+        }
+
+        result
+    }
+}
+
+struct StoredFileReader {
+    iter: Peekable<Box<dyn Iterator<Item = Result<Bytes>>>>,
+}
+
+impl StoredFileReader {
+    pub fn new(
+        stored_tree: Arc<StoredTree>,
+        entry: IndexEntry,
+        byte_offset: u64,
+        monitor: Arc<dyn Monitor>,
+    ) -> Result<Self> {
+        let file_content = entry
+            .addrs
+            .into_iter()
+            .scan(byte_offset, |skip_bytes, mut entry| {
+                if *skip_bytes == 0 {
+                    Some(entry)
+                } else if *skip_bytes < entry.len {
+                    entry.len -= *skip_bytes;
+                    entry.start += *skip_bytes;
+                    *skip_bytes = 0;
+                    Some(entry)
+                } else {
+                    *skip_bytes -= entry.len;
+                    None
+                }
+            })
+            .map::<Result<Bytes>, _>(move |entry| {
+                let content = stored_tree
+                    .block_dir()
+                    .get_block_content(&entry.hash, monitor.clone())?;
+
+                Ok(content.slice((entry.start as usize)..(entry.start + entry.len) as usize))
+            });
+
+        Ok(Self {
+            iter: (Box::new(file_content) as Box<dyn Iterator<Item = Result<Bytes>>>).peekable(),
+        })
+    }
+}
+
+impl Read for StoredFileReader {
```

Comment:
I think actually `io::Read` may not be especially a good interface for reading from our own sources, and I actually ripped it out of other places in d5a5a70f6b182f037c37804de5e92681d9567010. Maybe it's better to have an interface that returns an iter of Bytes.

Reply:
What exactly do you mean by ` for reading from our own sources`?